### PR TITLE
[TIP-702] Refactor behat entity creations

### PR DESCRIPTION
**Pull request by pierallard**
# What do this PR ?

## tl;dr;

It fixes entity behat creation to be as close as possible as our flat format and prevent any magic usage of entity creation + entity database check.
Don't be afraid of files number nor updated lines, 95% is only behats tables.

## Moar details

This PR:
- Uses the dedicated services (convertor / processor / writer) to convert, process and write entities in database
- Only do the refacto job for attributes, attribute groups, families, channels, product groups, variant groups, association type, group types
- Update "entity checkers" ("there should be the following ...") to raise an error when a column is unchecked, it's not the case on master.
- Update "entity checkers" to allow arrays checks regardless the order (`[a,b] === [b,a]`)
- Split "the following product groups" to "the following product groups" + "the following variant groups" because it's not the same entity services and same columns
- Because every of this changes has been done with a script, it also align tables in the behat.

## Want to review ?

You can easily only read the FixtureContext.php and forgot the .features IMHO.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | n
| Added Behats                      | n
| Added integration tests           | n
| Changelog updated                 | n
| Review and 2 GTM                  | Todo
| Migration script                  | -
| Tech Doc                          | -

### CODE REVIEW

File: features/Context/FixturesContext.php [hunk 1]
```diff
@@ -217,8 +225,20 @@ public function theFollowingAttributeGroups(TableNode $table)
...
+        $converter = $this->getContainer()->get('pim_connector.array_converter.flat_to_standard.attribute');
+        $processor = $this->getContainer()->get('pim_connector.processor.denormalization.attribute');
+        $saver     = $this->getContainer()->get('pim_catalog.saver.attribute');
+
         foreach ($table->getHash() as $data) {
-            $this->createAttribute($data);
+            $attribute = $processor->process($converter->convert($data));
+
+            if (isset($data['unique'])) {
+                // TODO Due to Pim/Component/Catalog/Updater/AttributeUpdater.php:226 (SDS-998)
```
Comment (grena) [inline features/Context/FixturesContext.php]: what is todo?

File: features/Context/FixturesContext.php [hunk 2]
```diff
@@ -178,8 +175,12 @@ public function anEnabledOrDisabledProduct($status, $sku)
      */
     public function theFollowingFamilies(TableNode $table)
     {
+        $converter = $this->getContainer()->get('pim_connector.array_converter.flat_to_standard.family');
+        $processor = $this->getContainer()->get('pim_connector.processor.denormalization.family');
+        $saver     = $this->getContainer()->get('pim_catalog.saver.family');
+
         foreach ($table->getHash() as $data) {
-            $this->createFamily($data);
+            $saver->save($processor->process($converter->convert($data)));
```
Comment (grena) [inline features/Context/FixturesContext.php]: no validation needed?

File: features/Context/FixturesContext.php [hunk 3]
```diff
@@ -413,27 +433,39 @@ public function createSelectAttributesWithOptions($attributeCount, $filterable,
...
+                    $requirements = [];
+                    foreach ($family->getAttributeRequirements() as $requirement) {
+                        if ($matches['channel'] === $requirement->getChannel()->getCode()) {
+                            $requirements[] = $requirement->getAttribute()->getCode();
+                        }
+                    }
+                    $this->assertArrayEquals(explode(',', $value), $requirements);
+
+                } else {
+                    throw new \InvalidArgumentException(sprintf('Can not check "%s" attribute of the family', $key));
```
Comment (grena) [inline features/Context/FixturesContext.php]: Cannot

File: features/Context/FixturesContext.php [hunk 4]
```diff
@@ -475,42 +507,36 @@ public function thereShouldBeTheFollowingChannels(TableNode $table)
...
-            assertEquals($data['currencies'], implode(',', $currencyCodes));
-
-            if ('' !== $data['conversion_units']) {
-                $units = explode(',', $data['conversion_units']);
-                $formattedUnits = [];
-                foreach ($units as $unit) {
-                    list($key, $value) = explode(':', trim($unit));
-                    $formattedUnits[trim($key)] = trim($value);
+                } else {
+                    throw new \InvalidArgumentException(sprintf('Can not check "%s" attribute of the channel', $key));
```
Comment (grena) [inline features/Context/FixturesContext.php]: Cannot

File: features/Context/FixturesContext.php [hunk 5]
```diff
@@ -2251,4 +2042,17 @@ protected function getDocumentManager()
...
+     * Assert than 2 arrays are equal, regardless the order of the elements.
+     *
+     * @param $array1
+     * @param $array2
+     */
+    protected function assertArrayEquals($array1, $array2)
+    {
+        sort($array1);
+        sort($array2);
+        assertEquals(join(', ', $array1), join(', ', $array2));
```
Comment (grena) [inline features/Context/FixturesContext.php]: `assertEquals` doesn't take order in account for arrays?

Comment (pierallard) [inline features/Context/FixturesContext.php]: Nope ! Example : assertEquals([1,2,3], [3,2,1]);
```     Failed asserting that two arrays are equal.
      --- Expected
      +++ Actual
      @@ @@
       Array (
      -    0 => 1
      +    0 => 3
           1 => 2
      -    2 => 3
      +    2 => 1
       )
```

File: features/Context/FixturesContext.php (hunk 2 above)
Comment (pierallard) [inline features/Context/FixturesContext.php]: Validation is done in the processor 
Pim/Component/Connector/Processor/Denormalization/Processor.php:72
It throws an InvalitItemException when line is not valid.

File: features/Context/FixturesContext.php (hunk 5 above)
Comment (grena) [inline features/Context/FixturesContext.php]: Ooooh ok, so it takes it in account, but we don't need it, I missread your code, thanks for the snippet 👍

File: features/Context/FixturesContext.php [hunk 6]
```diff
@@ -522,11 +546,25 @@ public function thereShouldBeTheFollowingChannels(TableNode $table)
      */
     public function thereShouldBeTheFollowingGroupTypes(TableNode $table)
     {
+        assertEquals([1,2,3], [3,2,1]);
```
Comment (anaelChardan) [inline features/Context/FixturesContext.php]: ????

File: features/Context/FixturesContext.php [hunk 7]
```diff
@@ -475,42 +505,36 @@ public function thereShouldBeTheFollowingChannels(TableNode $table)
...
         foreach ($table->getHash() as $data) {
             $channel = $this->getChannel($data['code']);
+            unset($data['code']);
 
-            if (isset($data['label-en_US'])) {
-                assertEquals($data['label-en_US'], $channel->getTranslation('en_US')->getLabel());
-            }
+            foreach ($data as $key => $value) {
+                $matches = null;
+                if ('tree' === $key) {
```
Comment (anaelChardan) [inline features/Context/FixturesContext.php]: Is there a way to split all these elseif conditions into checkers functions or other ? I don't know if it's really useful but I find this a little bit long (I know that it was already the case). :-)

File: features/Context/FixturesContext.php (hunk 5 above)
Comment (anaelChardan) [inline features/Context/FixturesContext.php]: +1 :-)

File: features/Context/FixturesContext.php (hunk 7 above)
Comment (pierallard) [inline features/Context/FixturesContext.php]: In fact, none of these functions are exactly the same. They "looks like" but in reality they don't have the same meaning.
That's 1 thing I recently learned: don't try to DRY everywhere.
